feat(content-item-horizontal-thumbnail): add thumbnail variation

mwanberg opened the pull request:
### Related Ticket(s)

DDS ticket: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6244
HC Jira ticket: https://jsw.ibm.com/browse/HC-1969


### Description

This PR adds a new variation to the "Content item horizontal" web component. It is very similar to the default pattern except: 

1) The eyebrow is removed
2) A smaller image has been added
3) On lg+ screens the layout goes to three columns, with the thumbnail on the right side.

![Screen Shot 2021-06-19 at 8 46 10 PM](https://user-images.githubusercontent.com/384598/122661456-673b0c80-d13f-11eb-8e7d-076047fa344f.png)


### Changelog

**New**

- N/A

**Changed**

- Added a "With Thumbnail" variation to `<dds-content-item-horizontal>`
- Small width style fix that affects the React `<dds-content-item-horizontal>` component

**Removed**

- N/A

### React component style fix

The current default variation for both the Web Component and React  `<dds-content-item-horizontal>` component has a small mis-alignment on md+ breakpoints. This PR introduces a fix for that, so we will see a small horizontal visual diff for the both the Web Component and React default variation.

**Current:**

![Screen Shot 2021-06-24 at 4 45 06 PM](https://user-images.githubusercontent.com/384598/123345837-d03dce00-d50b-11eb-9291-5fd5c06c650b.png)

**Fixed:**

![Screen Shot 2021-06-24 at 4 46 34 PM](https://user-images.githubusercontent.com/384598/123345858-db90f980-d50b-11eb-9ea7-5bc15ddf028c.png)


<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "package: styles": Carbon Expressive -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

annawen1 commented on packages/styles/scss/components/content-item-horizontal-thumbnail/_content-item-horizontal-thumbnail.scss:
```diff
@@ -0,0 +1,159 @@
+/**
```
don't think we need to have a separate `scss` file in the styles package, can you move this to the `scss` file you created within the `content-item-horizontal` web-components folder?

annawen1 commented on packages/styles/scss/components/content-item-horizontal-thumbnail/_content-item-horizontal-thumbnail.scss:
```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright IBM Corp. 2016, 2021
```
don't think we need to have a separate scss file in the styles package, can you move this to the scss file you created within the content-item-horizontal web-components folder? It's a little confusing to have 2 different files.

mwanberg replied:
@annawen1 I was using the `<dds-content-item-horizontal-media>` component/variation as an example, which had a setup similar to what I created. But I can move these styles over as you suggest!

mwanberg replied:
Also, I totally did make them duplicate. 🙈

annawen1 commented on packages/web-components/src/components/content-item-horizontal/__stories__/README.stories.mdx:
````diff
@@ -118,6 +118,34 @@ import '@carbon/ibmdotcom-web-components/es/components/content-item-horizontal/c
+
+```javascript
+import '@carbon/ibmdotcom-web-components/es/components/content-item-horizontal/index.js';
+import '@carbon/ibmdotcom-web-components/es/components/content-item-horizontal/content-item-horizontal-thumbnail';
````
since you've already import `content-item-horizontal-thumbnail` in `content-item-horizontal/index.js`, I don't think we need to add this import statement again in the readme.

annawen1 commented on packages/styles/scss/components/content-item-horizontal-thumbnail/_content-item-horizontal-thumbnail.scss:
```diff
@@ -0,0 +1,159 @@
+  }
+
+  :host(#{$dds-prefix}-content-item-horizontal) ::slotted([slot='heading']),
+  :host(#{$dds-prefix}-content-item-horizontal-copy) ::slotted(:not([slot])),
```
Haven't seen this `::slotted(:not([slot]))`, can you tell me what this selector is targeting?

mwanberg replied:
@annawen1 TBH that's a new one to me, too. I was once again copying what I was seeing already existing, e.g. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/0d0571e43b1732086681ac59f1f7193765b9d83b/packages/styles/scss/components/content-item-horizontal-media/_content-item-horizontal-media.scss#L30

annawen1 commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal-thumbnail.scss:
```diff
@@ -0,0 +1,148 @@
+    }
+  }
+
+  :host(#{$dds-prefix}-link-list) ::slotted([slot='footer']) {
```
I think this may be changing the layout of components using this element, can you contain it to just the `content-item-horizontal-thumbnail`?

annawen1 commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal-thumbnail.scss:
```diff
@@ -0,0 +1,148 @@
+    }
+  }
+
+  .#{$prefix}--content-item-horizontal-thumbnail__heading-wrapper {
```
instead of using a wrapper class, could you target the heading with the `:host(#{$dds-prefix}-content-item-horizontal-thumbnail) ::slotted([slot='heading'])` selector that you have below?

annawen1 reviewed (commented):
Thanks @mwanberg for getting this set up! Have a few comments, but otherwise it's looking good!

proeung commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal-thumbnail.ts:
```diff
@@ -0,0 +1,53 @@
+ * @element dds-content-item-horizontal-thumbnail
+ */
+@customElement(`${ddsPrefix}-content-item-horizontal-thumbnail`)
+class DDSContentItemHorizontalThumbnail extends DDSContentItem {
```
I don't think this new component `<dds-content-item-horizontal-thumbnail>` is needed for the work outlined in this issue (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/6110). It seems to me like we just need to include the [dds-mage](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/image/image.ts#L26) component within the [dds-content-item-horizontal](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/content-item-horizontal/content-item-horizontal.ts#L25) component and defining the `image` slot within the HTML return of the shadow dom, unless there's something else that I'm missing here.

@annawen1 @photodow Can you let me know if what I'm suggesting is in alignment with what you guys are thinking as well?

https://ibm.ent.box.com/file/812734380200

annawen1 replied:
@proeung that's a good point, from the functional specs I think we could work off of `content-item-horizontal` itself and just add an attribute to render the `thumbnail` styles

proeung commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal-thumbnail-copy.ts:
```diff
@@ -0,0 +1,38 @@
+ * @element dds-content-item-horizontal-thumbnail-copy
+ */
+@customElement(`${ddsPrefix}-content-item-horizontal-thumbnail-copy`)
+class DDSContentItemHorizontalThumbnailCopy extends StableSelectorMixin(DDSContentItemCopy) {
```
I don't think this component is needed as the `dds-content-item-horizontal-copy` has already existed and can be used within the variant of `content-item-horizontal` layout component with thumbnail. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/content-item-horizontal/content-item-horizontal-copy.ts

Similar comment as noted here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/6425/files#diff-962566a3c5ebb5bad8f6364e5cdadebedeb244446091770dc9b678a89b3f2743R25

proeung reviewed (changes requested):
@mwanberg This is a great first draft! You've definitely have put a lot of work and effort into this enhancement ticket, however, there are some implementation questions that I have surrounding extending some components to showcase a variant with a thumbnail and video player. I'll defer this to @annawen1 as perhaps there might be a good reason why extending these components ( eg. `dds-content-item-horizontal-thumbnail`, `dds-content-item-horizontal-thumbnail-copy`) are warranted.

mwanberg commented:
@annawen1 @proeung Based on your suggestions, I refactored, editing the default component instead of creating a new one. Please let me know if I'm on the right track.

I made changes directly to the original content item horizontal component, so I'm guessing we should check for any regressions. There will likely be one visible change, which is a slight adjustment in how wide the whole component is. There was a bug in the original code that made the width slightly less than what it should be. There should be no other visual changes for the default variation.

Because I needed to modify the default markup to accommodate the new thumbnail image and slightly different layout and margins of the new variation, there is some moderately significant changes in both the markup structure and the styles. I tried to keep things simple while changing as little of the original as possible, so it's possible the whole thing could use some more streamlining. I'm open to any observations or suggestions.

One problem I ran into is that I found it difficult to target the copy and the CTAs in the styling. I am still working to figure that out.

A few questions:
- Do I still need to do the new feature flag?
- Should I be testing this in RTL?
- Should I be writing any new tests?

pavolprok commented:
Hi there. Thanks for putting this component together. I had a look at it and found some issues:
1. On max, xlg, lg breakpoints the component starts on the 4th column from the left but it is suppose to start on the 5th column from the left. The 4:3 thumbnail mage is suppose to be placed on the last 3 right columns. 
![Screen Shot 2021-06-21 at 17 39 07](https://user-images.githubusercontent.com/70329414/122899797-6fb46280-d34c-11eb-9625-2454b6fd6c8a.png)

2. On the small breakpoint the 4:3 image is suppose to be left aligned and span max 3 columns.
![Screen Shot 2021-06-21 at 17 57 40](https://user-images.githubusercontent.com/70329414/122900295-d89bda80-d34c-11eb-816b-1363c0971d42.png)


3. I do not see the possibility to change the copy text in the knobs. There is only the possibility to change headline and CTAs.
![Screen Shot 2021-06-21 at 17 54 58](https://user-images.githubusercontent.com/70329414/122900024-a12d2e00-d34c-11eb-956c-3705247312ef.png)

annawen1 commented:
@mwanberg 

> There will likely be one visible change, which is a slight adjustment in how wide the whole component is. There was a bug in the original code that made the width slightly less than what it should be. There should be no other visual changes for the default variation.

That's a good catch!

> One problem I ran into is that I found it difficult to target the copy and the CTAs in the styling. I am still working to figure that out.

Depending on what you're trying to style, you may need to extend a component if necessary. Let me know if you need help with that.

>  A few questions:
> 
> - Do I still need to do the new feature flag?
> - Should I be testing this in RTL?
> - Should I be writing any new tests?

- I don't remember seeing that this needed to be behind a feature flag @shixiedesign can you confirm?
- We haven't actively been testing the components in RTL, but that's definitely something to consider. If you add the `RTL` label to this PR, it should build out a new deploy preview.
- Yes, could you take a look at other component tests we've written and follow those as a guide?

proeung commented on packages/web-components/src/components/content-item-horizontal/__stories__/content-item-horizontal.stories.ts:
```diff
@@ -63,6 +64,34 @@ export const Default = ({ parameters }) => {
+  const { alt, heading, copy, ctaType1, ctaCopy1, href1, ctaType2, ctaCopy2, href2 } =
+    parameters?.props?.ContentItemHorizontal ?? {};
+  return html`
+    <dds-content-item-horizontal thumbnail="true">
```
Where is this `thumbnail` attribute being defined? I'm not seeing this prop defined in the `dds-content-item-horizontal` component.

proeung commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal.ts:
```diff
@@ -26,12 +26,17 @@ class DDSContentItemHorizontal extends DDSContentItem {
         </div>
-        <div class="${prefix}--content-item-horizontal__col">
-          ${this._renderBody()}${this._renderFooter()}
+        <div class="${prefix}--content-item-horizontal__col ${prefix}--content-item-horizontal__col--2">
```
Thanks for defining a slot for the thumbnail/image in the HTML return. I think we'll need to add a condition check (eg. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/card/card.ts#L130) within the result that adds in the `div` wrapper when the `thumbnail` or `image` slot is defined within the markup of the `<dds-content-item-horizontal>`. @annawen1 Let me know if you agree.

Also, @shixiedesign correct me if I'm wrong, but we'll need to account for the `<dds-video-player>` component that opens up the lightbox media viewer (https://ibmdotcom-web-components.mybluemix.net/?path=/story/components-video-player--with-lightbox-media-viewer)? if not, feel free to disregard my comment.

annawen1 replied:
> Thanks for defining a slot for the thumbnail/image in the HTML return. I think we'll need to add a condition check (eg. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/card/card.ts#L130) within the result that adds in the div wrapper when the thumbnail or image slot is defined within the markup of the <dds-content-item-horizontal>. @annawen1 Let me know if you agree.

yep, same for the `eyebrow` slot. We can check if the `thumbnail` attribute exists/is true to render the `thumbnail` and `eyebrow` slots.

```
{this.thumbnail ? <slot name="eyebrow" @slotchange="${this._handleSlotChange}"></slot> : undefined}
```

annawen1 replied:
@pavolprok @shixiedesign want to clarify again if we need to account for `video` as well or will it only be `image` for thumbnail?

pavolprok replied:
@annawen1 the thumbnails will only be images. No videos. ;-)

proeung replied:
@pavolprok Thanks for the confirmation!

proeung commented on packages/web-components/src/components/content-item-horizontal/__stories__/README.stories.mdx:
```diff
@@ -131,6 +150,7 @@ See [our README](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/t
 | ------------------------------------- | -------------------------------------------------- | ----------- |
 | `<dds-content-item-horizontal>`       | `data-autoid="dds--content-item-horizontal"`       | Component   |
 | `<dds-content-item-horizontal-media>` | `data-autoid="dds--content-item-horizontal-media"` | Component   |
+| `<dds-content-item-horizontal-thumbnail>` | `data-autoid="dds--content-item-horizontal-thumbnail"` | Component   |
```
This line can be removed as we're no longer defining a component for `<dds-content-item-horizontal-thumbnail>`.

proeung commented on packages/web-components/src/components/content-item-horizontal/__stories__/README.stories.mdx:
```diff
@@ -60,6 +60,25 @@ import '@carbon/ibmdotcom-web-components/es/components/content-item-horizontal/i
+      Microservices and containers
+    </dds-link-list-item-cta>
+  </dds-link-list>
+  <dds-image slot="thumbnail" alt="Alt text" default-src="image.jpg"></dds-image>
```
```suggestion
  <dds-image slot="thumbnail" alt="Alt text" default-src="https://fpoimg.com/960x720?text=4:3&bg_color=ee5396&text_color=161616"></dds-image>
```

Let's use FPO image for easy copy and paste of this WC markup so that adopters can see what this variant will look like instead of defending the URL of the image.

proeung commented on packages/web-components/src/components/content-item-horizontal/__stories__/content-item-horizontal.stories.ts:
```diff
@@ -63,6 +64,34 @@ export const Default = ({ parameters }) => {
+          ${ctaCopy2}
+        </dds-link-list-item-cta>
+      </dds-link-list>
+      <dds-image slot="thumbnail" alt="${ifNonNull(alt)}" default-src="${imgMd4x3}"></dds-image>
```
@shixiedesign or @pavolprok Do we need to account for media queries for this thumbnail image? I'm not sure if there are different ratios or dimensions that we need to serve on different breakpoints.

pavolprok replied:
Hi Putra. There will be only 4:3 ratio for the thumbnail images. The thumbnail images should span across 3 last columns  on max, xlg, lg, md. The only chage it rems of placement of the images is on the small breakpoint where the images spans across 3 columns but is aligned to the left and is placed below the CTA.

proeung reviewed (changes requested):
@mwanberg Great revision! Thanks for applying the refactoring work that I've suggested. I just have a few comments, but overall looks good from a code perspective. @annawen1 or @IgnacioBecerra let me know if you guys have other feedback on this PR.

mwanberg commented:
@pavolprok Thank you for reviewing this PR! The branch has been updated with fixes from your feedback.

1. I have adjusted the component so it starts on the 5th column on lg+ and the thumbnail now is aligned right with the last column.
2. The thumbnail now fits inside the first three columns on mobile/sm.
3. After attempting to add a "copy" knob and then looking into it with @annawen1, it seems that there is currently an [open issue](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5648) where some "copy" knobs aren't updating properly. So unfortunately I think we will need to leave it out of this version.

shixiedesign reviewed (approved):
This looks good! Thank you for the contribution!

proeung reviewed (approved):
@mwanberg Code changes look good overall! Thanks for addressing my feedback!

@annawen1 @shixiedesign I'm approving this PR from a code review perspective, however, let us know if you see anything else that needs adjusting.

shixiedesign commented:
Hi! Almost approved - noticed a few things got affected: 

Looks like the changes are done to style package, and `content item horizontal -- default` in React is affected: 
**Current:** 
![image](https://user-images.githubusercontent.com/15144993/123327530-83a4c380-d508-11eb-8da7-3ae05a5b6b15.png)

**Expecting**
![image](https://user-images.githubusercontent.com/15144993/123327573-90291c00-d508-11eb-9cdc-9d993ae220b1.png)

mwanberg commented:
@shixiedesign Good catch! I've modified the styling to fix that regression. One thing that will still be different for both the Web Component and React default variation is a slight adjustment in the width of the component on md+ breakpoints. This fixes an existing bug. Please let me know if you'd like me to undo that fix.

**Current:**

![Screen Shot 2021-06-24 at 4 45 06 PM](https://user-images.githubusercontent.com/384598/123345837-d03dce00-d50b-11eb-9291-5fd5c06c650b.png)

**Fixed:**

![Screen Shot 2021-06-24 at 4 46 34 PM](https://user-images.githubusercontent.com/384598/123345858-db90f980-d50b-11eb-9ea7-5bc15ddf028c.png)


@proeung and @annawen1 With Anna's direction I updated a little of how the `thumbnail` property is working so now it's a little cleaner AND the eyebrow shows up now on the default variation.

What remains now is writing up tests.

pavolprok commented:
> Link to preview: https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/6425/index.html?path=/story/components-content-item-horizontal--with-thumbnail
> 
> @pavolprok Thank you for reviewing this PR! The branch has been updated with fixes from your feedback.
> 
> 1. I have adjusted the component so it starts on the 5th column on lg+ and the thumbnail now is aligned right with the last column.
> 
> ![Screen Shot 2021-06-24 at 11 06 38 AM](https://user-images.githubusercontent.com/384598/123311646-4296ba00-d4dc-11eb-9f0c-4d5daac12ffb.png)
> 
> 1. The thumbnail now fits inside the first three columns on mobile/sm.
> 
> ![Screen Shot 2021-06-24 at 11 05 04 AM](https://user-images.githubusercontent.com/384598/123311523-21ce6480-d4dc-11eb-9778-1bf4b288a035.png)
> 
> 1. After attempting to add a "copy" knob and then looking into it with @annawen1, it seems that there is currently an [open issue](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5648) where some "copy" knobs aren't updating properly. So unfortunately I think we will need to leave it out of this version.

Thank you for the updates. The component now looks great.

pavolprok reviewed (approved):
Thanks for addressing the alignment issues. The component now looks correct.

annawen1 commented on packages/web-components/src/components/content-item-horizontal/content-item-horizontal.ts:
```diff
@@ -23,16 +23,35 @@ const { stablePrefix: ddsPrefix } = ddsSettings;
-          ${this._renderBody()}${this._renderFooter()}
+        <div class="${prefix}--content-item-horizontal__col--1">
+          <div class="${prefix}--content-item-horizontal__heading-wrapper">
+            ${this.thumbnai
```
missing the `l` in `thumbnail` here

annawen1 reviewed (commented):
close! If we could also get the unit tests you've been working on, I think we should be good to go

shixiedesign commented:
React is fixed! Looks like 

1. web components's version's `content-item-horizontal: default` is still not showing eyebrow: 

**Current:** 
![image](https://user-images.githubusercontent.com/15144993/123430144-d5dff600-d595-11eb-85a5-2bc2e52ee60c.png)
**Expecting**
![image](https://user-images.githubusercontent.com/15144993/123430191-e2fce500-d595-11eb-8746-c925a55dc6e2.png)

2. this is totally not important but if it's easy, can we **nudge the `default` variant in storybook right by one column**? I think this is an existing error so out of scope of this PR. It's just so much more obvious now when the other variants are nice and correct.

shixiedesign reviewed (approved):
Woohoo!! And extra thanks for fixing the existing grid issue in the default component !! 🙏 🏅

annawen1 reviewed (approved):
Thanks @mwanberg! Looks good!
